processing/src/main/java/io/druid/query/extraction/ExplicitDimRenameFn.java

```diff
@@ -0,0 +1,129 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import io.druid.jackson.DefaultObjectMapper;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+/**
+ *
+ */
+public class ExplicitDimRenameFn extends NamespacedExtraction
+{
+  @JsonIgnore
+  private static final ObjectMapper mapper = new DefaultObjectMapper();
+
+  private final Map<String, String> renames;
+
+  @JsonCreator
+  public ExplicitDimRenameFn(
+      @NotNull @JsonProperty("renames")
+      final Map<String, String> renames,
+      @Nullable @JsonProperty("retainMissingValue")
+      final Boolean retainMissingValue,
+      @Nullable @JsonProperty("replaceMissingValueWith")
+      final String replaceMissingValueWith,
+      @Nullable @JsonProperty("uniqueRenames")
+      final Boolean uniqueRenames
+  )
+  {
+    super(
+        new Function<String, Function<String, String>>()
+        {
+          @Nullable
+          @Override
+          public Function<String, String> apply(@Nullable String input)
+          {
+            return new Function<String, String>(){
+              @Nullable
+              @Override
+              public String apply(@Nullable String input)
+              {
+                if (Strings.isNullOrEmpty(input)) {
+                  return null;
+                }
+                return renames.get(input);
+              }
+            };
+          }
+        },
+        "__explicitNamespace_druid_internal",
+        retainMissingValue,
+        replaceMissingValueWith,
+        uniqueRenames
+    );
+    this.renames = Preconditions.checkNotNull(renames, "`renames` required");
+  }
+
+  private static final byte CACHE_TYPE_ID = 0x5;
+
+  @JsonProperty("renames")
+  public Map<String, String> getRenames()
+  {
+    return renames;
+  }
+
+  @Override
+  @JsonProperty("retainMissingValue")
+  public boolean isRetainMissingValue() {return super.isRetainMissingValue();}
+
+  @Override
+  @JsonProperty("replaceMissingValueWith")
+  public String getReplaceMissingValueWith() {return super.getReplaceMissingValueWith();}
+
+  @Override
+  @JsonProperty("uniqueRenames")
+  public boolean isUniqueRenames()
+  {
+    return super.isUniqueRenames();
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    final byte[] bytes;
+    try {
+      bytes = mapper.writeValueAsBytes(renames);
```

Comment:
using the JSON representation is actually gonna be pretty not great for space efficiency.  It would probably be better to iterate through the entries and manually concatenate them.